models/userStatus.js

```diff
@@ -327,101 +327,111 @@ const updateStatusOnTaskCompletion = async (userId) => {
-        latestStatusData = await getUserStatus(userId);
-      } catch (error) {
-        usersWithStatusUnchanged++;
-        return batch;
-      }
-      const { id, userStatusExists, data } = latestStatusData;
+  const summary = {
+    totalUsers: users.length,
+    totalUnprocessedUsers: 0,
+    totalOnboardingUsersAltered: 0,
+    totalOnboardingUsersUnAltered: 0,
+    totalActiveUsersAltered: 0,
+    totalActiveUsersUnAltered: 0,
+    totalIdleUsersAltered: 0,
+    totalIdleUsersUnAltered: 0,
+  };
 
-      if (!userStatusExists || !data?.currentStatus) {
-        const newUserStatusRef = userStatusModel.doc();
-        const newUserStatusData = {
-          userId,
-          currentStatus: {
-            state: userState.IDLE,
-            message: "",
-            from: currentTimeStamp,
-            until: "",
-            updatedAt: currentTimeStamp,
-          },
-        };
-        batch.set(newUserStatusRef, newUserStatusData);
-      } else {
-        const {
-          currentStatus: { state, until },
-        } = data;
+  for (const { userId, state } of users) {
+    let latestStatusData;
+    try {
+      latestStatusData = await getUserStatus(userId);
+    } catch (error) {
+      summary.totalUnprocessedUsers++;
+      continue;
+    }
+    const { id, userStatusExists, data } = latestStatusData;
+    const statusToUpdate = {
+      state,
+      message: "",
+      from: currentTimeStamp,
+      until: "",
+      updatedAt: currentTimeStamp,
+    };
 
-        if (state === userState.OOO || state === userState.ACTIVE) {
-          const docRef = userStatusModel.doc(id);
-          const updatedStatusData =
-            state === userState.OOO
-              ? {
-                  futureStatus: {
-                    state: userState.IDLE,
-                    message: "",
-                    from: until,
-                    until: "",
-                    updatedAt: currentTimeStamp,
-                  },
-                }
-              : {
-                  currentStatus: {
-                    state: userState.IDLE,
-                    message: "",
-                    from: currentTimeStamp,
-                    until: "",
-                    updatedAt: currentTimeStamp,
-                  },
-                };
+    if (!userStatusExists || !data?.currentStatus) {
+      const newUserStatusRef = userStatusModel.doc();
+      const newUserStatusData = {
+        userId,
+        currentStatus: statusToUpdate,
+      };
+      state === userState.ACTIVE ? summary.totalActiveUsersAltered++ : summary.totalIdleUsersAltered++;
+      batch.set(newUserStatusRef, newUserStatusData);
+    } else {
+      const {
+        currentStatus: { state: currentState, until },
+      } = data;
+      if (currentState === state) {
+        currentState === userState.ACTIVE ? summary.totalActiveUsersUnAltered++ : summary.totalIdleUsersUnAltered++;
+        continue;
+      }
+      if (currentState === userState.ONBOARDING) {
```

Comment:
here if the current status of the user is ONBOARDING. However, the currentStatus property of the data object is only defined if the user already has a user status. If the user does not have a user status, then the currentStatus property will be undefined.

This means that the if statement will always be true if the user does not have a user status. This will cause the code to update the user status of the user, even if the user does not have a user status.

Reply:
> here if the current status of the user is ONBOARDING. However, the currentStatus property of the data object is only defined if the user already has a user status. If the user does not have a user status, then the currentStatus property will be undefined.
> 
> This means that the if statement will always be true if the user does not have a user status. This will cause the code to update the user status of the user, even if the user does not have a user status.

`here if the current status of the user is ONBOARDING. However, the currentStatus property of the data object is only defined if the user already has a user status. If the user does not have a user status` - This statement is self contradictory. If the user is onboarding then `user does not have a user status.` cant be true. The current status can be either onboarding , idle , active or OOO. currently any user signing up for rds their status gets created automatically and assigned onboarding.